### PR TITLE
Fix: Pass **kwargs to infer.query calls in CausalInference.query

### LINKED ISSUE #2401: [BUG] "query" function in CausalInference.py ignores **kwargs
**Describe the bug**

The "query" function in CausalInference.py can use VariableElimination or  BeliefPropagation algorithms. Both those algorithms can handle virtual evidence natively. However, passing virtual_evidence as a **kwarg in the "query" function within CausalInference does nothing and is ignored completely.

**To Reproduce**

You can either pass a simple virtual_evidence CPD in a toy network and watch it get ignored or you can check the code and see that **kwargs are not passed to any of the "infer.query" calls in the function (which is where virtual_evidence for example would be passed to VariableElimination or BeliefPropagation algorithms).

**Expected behavior**
CausalInference.query should be able to pass **kwargs to the inference algorithms.

**Versions**

pgmpy==1.0.0
python==3.13.2

### PR DESCRIPTION
**Pull request by MAVRICK-1**
This pull request enhances the flexibility of the `CausalInference.query` method by allowing additional keyword arguments to be passed to underlying inference queries, and introduces a new test to verify support for virtual evidence. The most important changes are grouped below:

**CausalInference interface improvements:**

* The `query` method in `pgmpy/inference/CausalInference.py` now forwards any extra keyword arguments (`**kwargs`) to the underlying `infer.query` calls, enabling support for features such as virtual evidence and future extensibility. [[1]](diffhunk://#diff-cd91f402ff6e5bd677be1cc926967bde8a492f5d4d63ec687dbfa40236210d68L1079-R1084) [[2]](diffhunk://#diff-cd91f402ff6e5bd677be1cc926967bde8a492f5d4d63ec687dbfa40236210d68L1098-R1098) [[3]](diffhunk://#diff-cd91f402ff6e5bd677be1cc926967bde8a492f5d4d63ec687dbfa40236210d68L1117-R1119) [[4]](diffhunk://#diff-cd91f402ff6e5bd677be1cc926967bde8a492f5d4d63ec687dbfa40236210d68L1136-R1138)

**Testing and validation:**

* Added a new test class `TestQueryWithVirtualEvidence` in `pgmpy/tests/test_inference/test_CausalInference.py` to verify that the updated `query` method correctly handles queries with virtual evidence, confirming expected probability calculations.

image

Fixes https://github.com/pgmpy/pgmpy/issues/2401

### CODE REVIEW

**Comment by MAVRICK-1** [comment]
cc @ankurankan

File: pgmpy/inference/CausalInference.py [hunk 1]
```diff
@@ -1095,7 +1095,7 @@ def query(
             if var in adjustment_set.intersection(evidence.keys())
         }
         if len(evidence_adj_inter) != 0:
-            p_z = infer.query(adjustment_set, show_progress=False).reduce(
+            p_z = infer.query(adjustment_set, show_progress=False, **kwargs).reduce(
```
Comment (ankurankan) [inline pgmpy/inference/CausalInference.py]: I don't think we can pass `**kwargs` here as we actually need to compute `P(Z=z)` here.

File: pgmpy/inference/CausalInference.py [hunk 2]
```diff
@@ -1114,7 +1114,9 @@ def query(
                     },
                 )
         else:
-            p_z = infer.query(adjustment_set, evidence=evidence, show_progress=False)
+            p_z = infer.query(
```
Comment (ankurankan) [inline pgmpy/inference/CausalInference.py]: Same here.

File: pgmpy/tests/test_inference/test_CausalInference.py [hunk 3]
```diff
@@ -1343,3 +1343,24 @@ def test_estimate_multiple_paths(self):
         self.assertAlmostEqual(
             infer.estimate_ate("X", "Y", data), ((0.8 * 0.9) + (0.9 * 0.1)), places=1
         )
+
+
+class TestQueryWithVirtualEvidence(unittest.TestCase):
```
Comment (ankurankan) [inline pgmpy/tests/test_inference/test_CausalInference.py:1348]: I think we should consider a slightly more complicated case to test some edge cases. I think a possible model could be something that has at least one confounding variable and a collider variable. Maybe something like: `A -> B  A <- X -> B   A -> C <- B`.

**Review by ankurankan — CHANGES_REQUESTED** [review]
@MAVRICK-1 Thanks a lot for the PR. I think we need to be a bit careful here to make sure that we are computing the correct values as the differences are quite subtle. I think having a more thorough test case would really help in making sure that our implementation is correct.

File: pgmpy/tests/test_inference/test_CausalInference.py (hunk 3 above)
Comment (MAVRICK-1) [reply pgmpy/tests/test_inference/test_CausalInference.py:1348]: added new test class TestQueryWithConfoundingAndCollider which  includes:

-      A model with the structure A -> B, A <- X -> B, A -> C <- B
-     Tests with virtual evidence on confounder X
-     Tests with collider in the adjustment set
-     Tests with confounder in the adjustment set

**Comment by MAVRICK-1** [comment]
Cc @ankurankan

File: pgmpy/tests/test_inference/test_CausalInference.py [hunk 4]
```diff
@@ -1343,3 +1343,98 @@ def test_estimate_multiple_paths(self):
...
+        # Test with collider in the adjustment set
+        # When C is in the adjustment set, it should open the path X -> A -> C <- B
+        # which would introduce bias, but our algorithm should handle this correctly
+        query_result_with_adjustment = self.inference.query(
+            variables=["B"],
+            do={"A": 1},
+            adjustment_set={"C"},
+            virtual_evidence=virtual_evidence,
+        )
+
```
Comment (ankurankan) [inline pgmpy/tests/test_inference/test_CausalInference.py:1427]: I think we will need to manually compute values for these and check that the method is returning correct values.

**Review by ankurankan — CHANGES_REQUESTED** [review]
@MAVRICK-1 Please refer comment. Let me know if you need help in computing the values.

**Comment by MAVRICK-1** [comment]
on  it

**Comment by ankurankan** [comment]
Closing due to inactivity.

**Comment by MAVRICK-1** [comment]
@ankurankan can you please  open it my exam are going on will work on this and rest after 19th march

**Comment by ankurankan** [comment]
@MAVRICK-1 Feel free to reopen it when you start working on it again.
